streams/src/main/java/org/apache/kafka/streams/state/Stores.java

```diff
@@ -0,0 +1,206 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.common.utils.SystemTime;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+/**
+ * Factory for creating key-value stores.
+ */
+public final class Stores {
+
+    /**
+     * Begin to create a key-value store that keeps all entries in-memory for fast access, but that records changes to a Kafka
+     * topic so the store can be recovered.
+     * 
+     * @return the factory to create in-memory key-value stores; never null
+     */
+    public static InMemoryFactory inMemory() {
+        return new InMemoryKeyValueStoreFactory();
+    }
+
+    /**
+     * Begin to create a key-value store that stores all entries off-heap in a local database, but that records changes to a Kafka
+     * topic so the store can be recovered.
+     * 
+     * @return the factory to create off-heap key-value stores; never null
+     */
+    public static OffHeapFactory offHeap() {
+        return new RocksDBKeyValueStoreFactory();
+    }
+
+    /**
+     * The factory for creating any kind of key-value store.
+     */
+    public static interface KeyValueStoreFactory {
+
+        /**
+         * Use the given {@link Time} instance. By default, the {@link SystemTime} is used.
+         * 
+         * @param time the time provider; may be null if the system time should be used
+         * @return this factory instance
+         * @see #withSystemTime()
+         */
+        public KeyValueStoreFactory withTime(Time time);
+
+        /**
+         * Use the {@link SystemTime} for the store.
+         * 
+         * @return this factory instance
+         * @see #withTime(Time)
+         */
+        public KeyValueStoreFactory withSystemTime();
```

Comment:
Time object is only used for recording metrics, and hence users who do not create their own store implementation (i.e. the first category of users) should not be worry about what time to use; in addition other timers other than system time are used usually for tests, so I'm wondering if we want to expose those in the public APIs for the end users?